### PR TITLE
GitLab Delegated Joining

### LINKED ISSUE #17059: Machine ID: Support delegated joining on GitLab
Reference GitHub implementation at https://github.com/gravitational/teleport/pull/16938

https://docs.gitlab.com/ee/ci/cloud_services/

https://docs.gitlab.com/ee/ci/variables/index.html#pass-an-environment-variable-to-another-job

### Information available in GitLab tokens

CI_JOB_JWT:
```json
{
  "alg": "RS256",
  "kid": "4i3sFE7sxqNPOT7FdvcGA1ZVGGI_r-tsDXnEuYT4ZqE",
  "typ": "JWT"
},
{
  "namespace_id": "1008548",
  "namespace_path": "strideynet",
  "project_id": "39980849",
  "project_path": "strideynet/test-gitlab-jwt",
  "user_id": "842508",
  "user_login": "strideynet",
  "user_email": "redacted",
  "pipeline_id": "659183071",
  "pipeline_source": "push",
  "job_id": "3131574536",
  "ref": "main",
  "ref_type": "branch",
  "ref_protected": "true",
  "jti": "03362687-b43c-404a-b3bd-0a57e2f7b512",
  "iss": "gitlab.com",
  "iat": 1664985077,
  "nbf": 1664985072,
  "exp": 1664988677,
  "sub": "job_3131574536"
}
```

CI_JOB_JWT_V2
```json
{
  "alg": "RS256",
  "kid": "4i3sFE7sxqNPOT7FdvcGA1ZVGGI_r-tsDXnEuYT4ZqE",
  "typ": "JWT"
},
{
  "namespace_id": "1008548",
  "namespace_path": "strideynet",
  "project_id": "39980849",
  "project_path": "strideynet/test-gitlab-jwt",
  "user_id": "842508",
  "user_login": "strideynet",
  "user_email": "redacted",
  "pipeline_id": "659183071",
  "pipeline_source": "push",
  "job_id": "3131574536",
  "ref": "main",
  "ref_type": "branch",
  "ref_protected": "true",
  "jti": "f5e60e65-c5c8-4a27-a4d9-e689d580d084",
  "iss": "https://gitlab.com",
  "iat": 1664985077,
  "nbf": 1664985072,
  "exp": 1664988677,
  "sub": "project_path:strideynet/test-gitlab-jwt:ref_type:branch:ref:main",
  "aud": "https://gitlab.com"
}
```

### PR DESCRIPTION
**Pull request by strideynet**
Closes https://github.com/gravitational/teleport/issues/17059

- [x] Automated tests
- [x] Manual test GitLab cloud
- [x] Manual test GitLab self-hosted

Docs will come in a future PR o7

### CODE REVIEW

File: api/proto/teleport/legacy/types/types.proto [hunk 1]
```diff
@@ -1182,6 +1184,24 @@ message ProvisionTokenSpecV2GitHub {
   string EnterpriseServerHost = 2 [(gogoproto.jsontag) = "enterprise_server_host,omitempty"];
 }
 
+// ProvisionTokenSpecV2GitLab contains the GitLab-specific part of the
+// ProvisionTokenSpecV2
+message ProvisionTokenSpecV2GitLab {
+  message Rule {
+    // Sub also known as Subject is a string that roughly uniquely identifies
+    // the workload. The format of this varies depending on the type of
+    // github action run.
```
Comment (zmb3) [inline api/proto/teleport/legacy/types/types.proto:1193]:
```suggestion
    // GitLab action run.
```

Comment (strideynet) [reply api/proto/teleport/legacy/types/types.proto]: 03e13a0d8e

**Comment by strideynet** [comment]
image

As of 6c85c3f347 - we now have working delegated joining. All that remains is adding all the remaining allow rule options and testing against self-hosted GitLab.

**Comment by strideynet** [comment]
Remaining task is to stand up a GitLab instance and test against this instance !

**Comment by strideynet** [comment]
GitLab instance stood up - will hopefully run a manual test against this instance tomorrow morning.

**Comment by strideynet** [comment]
## Self-hosted test

image

**`.gitlab-ci.yml`**
```yaml
stages:          # List of stages for jobs, and their order of execution
  - build

build-job:       # This job runs in the build stage, which runs first.
  stage: build
  id_tokens:
    TBOT_GITLAB_JWT:
      aud: teleport.example.com
  script:
    - ./tbot start --token=gitlab-self-hosted-bot --destination-dir=./tbot-user --data-dir=./tbot-data --auth-server=teleport.example.com:443 --join-method=gitlab --oneshot
    - TELEPORT_LOGIN=root ./tsh -i ./tbot-user/identity --proxy teleport.example.com:443 ls
```
**ProvisionToken resource**
```yaml
kind: token
version: v2
metadata:
  name: gitlab-self-hosted-bot
  expires: "3000-01-01T00:00:00Z"
spec:
  roles: [Bot]
  join_method: gitlab
  bot_name: gitlab-demo
  gitlab:
    domain: gitlab.example.com
    allow:
      - project_path: noah/gitlab-sandbox
        ref_type: branch
        ref: main
        namespace_path: noah
```

I included some additional fields more than necessary in the allow rules just to check that all of these function as expected.

**Comment by strideynet** [comment]
Also manually tested:
- Incorrect audience token is rejected
- Correct issuer is enforced based on whether or not `domain` is set (e.g a Cloud issued ID token cannot be used against a self-hosted GitLab provision token)

**Review by zmb3 — APPROVED** [review]

**Review by hatched — APPROVED** [review]

**Pull request merged**